### PR TITLE
Insert sequences for dynamical decoupling

### LINKED ISSUE #1189: Implement `insert_ddd_sequences()` function for dynamical decoupling
Implement the `insert_ddd_sequences()` function described in the [RFC for dynamical decoupling](https://docs.google.com/document/d/1cRwFCTn6kUjI1P0kNydtevxIYtE4r8Omd_iWK0Pe8qo/edit?usp=sharing).

This function serves to fill slack windows of idling qubits with the optionally provided dynamical decoupling rule, and defaults to the random rule (i.e. a semi-random sequence of spin echoes).

An implementation skeleton with pseudo-code is present in the [RFC](https://docs.google.com/document/d/1cRwFCTn6kUjI1P0kNydtevxIYtE4r8Omd_iWK0Pe8qo/edit?usp=sharing) but, of course, a different approach can be used.

### PR DESCRIPTION
**Pull request by Aaron-Robertson**
Description
-----------

Fixes #1189. Adds `insert_ddd_sequences()` to `mitiq.ddd`.

License
-------

- [x] I license this contribution under the terms of the GNU GPL, version 3 and grant Unitary Fund the right to provide additional permissions as described in section 7 of the GNU GPL, version 3.

### CODE REVIEW

**Comment by Aaron-Robertson** [comment]
@andreamari I realized an oversight when applying rules recursively (as in some of our example cases), because we weren't initially checking for `I` gates in our slack constructor. Furthermore, if the moments are filled already, they need to be replaced (see `_construct_replacements()`). Not sure how we want to handle the fill/emptiness dynamically though, I'll reach out on discord.

File: mitiq/ddd/insertion.py [hunk 1]
```diff
@@ -80,3 +82,52 @@ def get_slack_matrix_from_circuit_mask(mask: np.ndarray) -> np.ndarray:
...
+    circuit: Circuit, sequence: Circuit, index: int
+) -> List[Tuple[int, Operation, Operation]]:
+    """Returns the replacements for insert_ddd_sequences."""
+
+    return [
+        (index + idx, old_moment.operations, new_moment.operations[0])
+        for new_moment, (idx, old_moment) in zip(
+            sequence, enumerate(circuit[index : index + len(sequence)])
+        )
+    ]
```
Comment (andreamari) [inline mitiq/ddd/insertion.py:97]: We probably don't need this function based on the current implementation of `insert_ddd_sequences`.

Comment (andreamari) [inline mitiq/ddd/insertion.py:97]: We probably don't need this function based on the current implementation of `insert_ddd_sequences`.

File: mitiq/ddd/insertion.py [hunk 2]
```diff
@@ -83,3 +85,40 @@ def get_slack_matrix_from_circuit_mask(mask: np.ndarray) -> np.ndarray:
...
+    circuit: Circuit,
+    rule: Callable[[int], Circuit],
+) -> Circuit:
+    """Returns the circuit with DDD sequences applied according to the input rule.
+
+    Args:
+        circuit: QPROGRAM circuit to execute with DDD.
+        rule: DDD rule function to apply.
+
+    Returns: Reconverted Circuit with spin echoes added.
```
Comment (andreamari) [inline mitiq/ddd/insertion.py:101]:
```suggestion
    """Returns the circuit with DDD sequences applied according to the input rule.

    Args:
        circuit: The QPROGRAM circuit to be modified with DDD sequences.
        rule: The rule determining what DDD sequences should be applied.
            A set of built-in DDD rules can be imported from ``mitiq.ddd.rules``.

    Returns: 
        The circuit with DDD sequences added.
```

File: mitiq/ddd/insertion.py [hunk 3]
```diff
@@ -83,3 +85,40 @@ def get_slack_matrix_from_circuit_mask(mask: np.ndarray) -> np.ndarray:
...
+                    {LineQubit(0): qubits[row_index]}
+                )
+                for idx, op in enumerate(ddd_sequence.all_operations()):
+                    circuit_with_ddd._moments[
+                        moment_idx + idx
+                    ] = circuit_with_ddd._moments[
+                        moment_idx + idx
+                    ].with_operations(
+                        op
+                    )
```
Comment (andreamari) [inline mitiq/ddd/insertion.py:123]: Very clean way of doing it!

**Review by andreamari — APPROVED** [review]
Thanks @Aaron-Robertson! Looks great. 
See comment on docstring. If applied, it may require some style fixing.

File: mitiq/ddd/insertion.py [hunk 4]
```diff
@@ -83,3 +85,40 @@ def get_slack_matrix_from_circuit_mask(mask: np.ndarray) -> np.ndarray:
                         break
 
     return slack_matrix
+
+
+@noise_scaling_converter
+def insert_ddd_sequences(
+    circuit: Circuit,
```
Comment (andreamari) [inline mitiq/ddd/insertion.py:92]:
```suggestion
    circuit: QPROGRAM,
```
I would suggest this change. I know that in other places we use `Circuit`, but `QPROGRAM` is probably more appropriate.

**Comment by andreamari** [comment]
The docs problem is unrelated to this PR and is fixed in #1251

File: mitiq/ddd/insertion.py (hunk 4 above)
Comment (Aaron-Robertson) [reply mitiq/ddd/insertion.py:109]: Ah, fair point. I added a second function that takes `QPROGRAM` (with docstring) and is external, with the existing function made internal (but still using the noise scaling converter). How's that?

Comment (andreamari) [reply mitiq/ddd/insertion.py:111]: Sounds good!

**Comment by andreamari** [comment]
@Aaron-Robertson, LGTM! I leave you the pleasure of merging when all tests are green! 
Please make sure to use _Squash and merge_ instead of _merge_. In this way the commit history is cleaner.

**Pull request merged**